File: src/editor/database/symbol/FunctionalSymbol.java
package editor.database.symbol;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * This class represents a symbol that has a specific meaning that isn't related to mana costs.
 *
 * @author Alec Roelke
 */
public class FunctionalSymbol extends Symbol
{
    /**
     * The Chaos symbol that appears on Plane cards.  Because it has a special
     * text representation that doens't use {}, it is exposed as a separate constant.
     */
    public static final FunctionalSymbol CHAOS = new FunctionalSymbol("chaos.png", "CHAOS")
    {
        @Override
        public String toString()
        {
            return "CHAOS";
        }
    };

    /**
     * Map of symbol texts onto their respective symbols.
     */
    public static final Map<String, FunctionalSymbol> SYMBOLS;

    static
    {
        Map<String, FunctionalSymbol> symbols = new HashMap<>();
        // Chaos symbol.
        symbols.put(CHAOS.toString(), CHAOS);
        // Phyrexian phi symbol.
        symbols.put("P", new FunctionalSymbol("phyrexia.png", "P"));
        // Tap symbol.  Used in costs in card text.
        symbols.put("T", new FunctionalSymbol("tap.png", "T"));
        symbols.put("TAP", symbols.get("T"));
        // Untap symbol.  Used in costs in card text.
        symbols.put("Q", new FunctionalSymbol("untap.png", "Q"));
        // Energy counter symbol.
        symbols.put("E", new FunctionalSymbol("energy.png", "E"));

        SYMBOLS = Collections.unmodifiableMap(symbols);
    }

    /**
     * Get the FunctionalSymbol corresponding to the given String.
     *
     * @param s String to look up
     * @return the FunctionalSymbol corresponding to the given String
     * @throws IllegalArgumentException if the String doesn't correspond to a functional symbol
     */
    public static FunctionalSymbol parseFunctionalSymbol(String s) throws IllegalArgumentException
    {
        FunctionalSymbol symbol = tryParseFunctionalSymbol(s);
        if (symbol == null)
            throw new IllegalArgumentException('"' + s + "\" is not a functional symbol");
        return symbol;
    }

    /**
     * Get the FunctionalSymbol corresponding to the given String.
     *
     * @param s String to look up
     * @return the FunctionalSymbol corresponding to the given String, or null if there is none
     */
    public static FunctionalSymbol tryParseFunctionalSymbol(String s)
    {
        return SYMBOLS.get(s.toUpperCase());
    }

    /**
     * Create a new FunctionalSymbol.
     *
     * @param iconName name of the new symbol's icon
     * @param text     text representation of the new symbol
     */
    private FunctionalSymbol(String iconName, String text)
    {
        super(iconName, text);
    }
}

File: src/main/scala/editor/database/symbol/FunctionalSymbol.scala
package editor.database.symbol

/**
 * Companion to [[FunctionalSymbol]] that contains all the existing non-mana symbols and functions for parsing strings
 * into [[FunctionalSymbol]]s. The possible [[FunctionalSymbol]]s are:
 * - The Chaos symbol
 * - The Phyrexia symbol
 * - The tap symbol
 * - The untap symbol
 * - The energy symbol
 *
 * @author Alec Roelke
 */
object FunctionalSymbol extends SymbolParser[FunctionalSymbol] with HasSymbolValues[String, FunctionalSymbol] {
  /** The Chaos symbol used on Planes, which has a special string representation without braces. */
  val Chaos = new FunctionalSymbol("chaos", "CHAOS") { override val toString = "CHAOS" }

  override val values = Map(
    Chaos.toString -> Chaos,
    "PW" -> Chaos,
    "P" -> new FunctionalSymbol("phyrexia", "P"),
    "T" -> new FunctionalSymbol("tap", "T"),
    "TAP" -> new FunctionalSymbol("tap", "T"),
    "Q" -> new FunctionalSymbol("untap", "Q"),
    "E" -> new FunctionalSymbol("energy", "E")
  )

  override def parse(s: String) = values.get(s.toUpperCase)
}

/**
 * A symbol with a specific meaning that's not used in mana costs.
 * 
 * @constructor create a new non-mana symbol
 * @param icon icon for the symbol
 * @param text text representation of the symbol
 * 
 * @author Alec Roelke
 */
class FunctionalSymbol private(icon: String, text: String) extends Symbol(icon, text)